cultureamp/kaizen-design-system
# feat: Secondary Single Select Reversed (React)

![image](https://user-images.githubusercontent.com/702885/85506324-7920c700-b633-11ea-9d4a-8703db934189.png)


Branch preview:
https://dev.cultureamp.design/dst/secondary-select-react/storybook-static/?path=/story/select-react--single-secondary-with-ellipsis

Figma:
https://www.figma.com/file/GMxm8rvDCbj0Xw3TQWBZ8b/Zen-UI-Kit?node-id=1929%3A25645



### In scope
- Secondary Single Select Reversed variant
- Visually meet the Zen design 80%, but not full Zen uplift (see out of scope)
- separate prop: `selectControlWidth`: `fillContainer` | `containSelection`. The default variant defaults to `fillContainer` and the secondary variant defaults to `containSelection`. When `containSelection` is true, the size of the "control" is shrink wrapped to the selected option text .
- truncate with ellipsis when the size of the control is bigger than a fixed width container

### Out of scope (these can be followed up in separate PRs):
- Secondary Single Select Non-Reversed (white) variant (should be easy, but descoping in the interest of scope hammering)
- the label (the current component doesn't support a label - it's currently BYO label)
- Full Zen uplift of related variants and where code (css) will shared between variants (eg: new focus rings). See https://github.com/cultureamp/kaizen-design-system/pull/548#issuecomment-647857515 for more details.

- Accessibility fixes to existing Select component (if there are any)

- this pre-existing bug in the searchable variant:
![image](https://user-images.githubusercontent.com/702885/85481643-8fb02980-b605-11ea-8349-0a2565469619.png)
- Animation (possibly the control size should animate when it suddenly changes size to make it less jarring?)
- Searchable variant
- Secondary Multi Select and other more complicated versions
- Elm versions

### TODO
- [x] reversed secondary props and styling
- [x] `selectControlWidth` prop and basic implementation for `secondary` (not implemented for default)
- [x] `containSelection` should ellipsize when the control is wider than a fixed width container 
- [x] `selectControlWidth` should default to `containSelection` for `secondary`
- [x] make stories that exercise edge cases
- [ ] TS docs for storybook

### TBC

- [x] naming things
- [x] should the control we the width of the selected option, or be the width of the widest option, like a native `select`?

### Notes for the Code Reviewer

- I tried to add JS Docs, but they don't seem to be appearing in Storybook. Any tips to get this working would be appreciated? At least the docs exist in the source code for now.
- is `variant` the best prop name? Is something like `prominence` better? Acording to the Kaizen docs (https://cultureamp.design/components/select/), variant is `single` or `multi`, but that's not true in the code API, as the API is inherited from `react-select`.

### QA
- make sure to thoroughly regression test all of the other existing Select component variations, both in Chromatic and Storybook(functional). Shall we get Dylan? I feel we could make better use of our QA staff
- regression test the Elm select (I don't think it uses any shared style, but just in case)

### Design review screenshots (IE11)
Ignore the missing pixels: that is just because of the emulator (graphics card emulation issue?)
![image](https://user-images.githubusercontent.com/702885/85504246-f39b1800-b62e-11ea-8e5a-ebf23c6ec3db.png)
![image](https://user-images.githubusercontent.com/702885/85504254-fac22600-b62e-11ea-8ef7-b92380c1fbdc.png)
![image](https://user-images.githubusercontent.com/702885/85504277-04e42480-b62f-11ea-8430-5559392e2a7f.png)

Just use Chromatic/Storybook for Non IE11
